docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,142 @@
+Pluggable Allocators
+=====
+
+Open Enclave provides a default memory allocator for enclaves.
+This default allocator, [dlmalloc (Doug Lea Malloc)](http://gee.cs.oswego.edu/dl/html/malloc.html), is not performant in multi-thread applications.
+
+There are various high-performance memory allocators that try to optimize, among other things, multi-thread scenarios:
+- [The GNU Allocator](https://www.gnu.org/software/libc/manual/html_node/The-GNU-Allocator.html) which is based on `pthreads malloc`, which is in turn based on `dlmalloc`.
+- [Thread-Caching Malloc (tcmalloc)](https://gperftools.github.io/gperftools/tcmalloc.html) developed by Google.
+- [snmalloc](https://github.com/Microsoft/snmalloc) developed by Micrsoft Research.
+- and many more. Application specific allocators often outperform general-purpose allocators.
+
+
+This document describes the design of a scheme to allow custom allocators to be used in enclaves.
+
+
+Precedents
+------
+
+- The GNU C Library provides a way to replace its malloc implementation, documented in [Replacing Malloc](https://www.gnu.org/software/libc/manual/html_node/Replacing-malloc.html#Replacing-malloc). </br>
+  To replace the allocator, the following replacement functions must be provided `malloc`, `free`, `calloc`, `realloc`, `aligned_alloc`, `malloc_usable_size`, `memalign`, `posix_memalign`, `palloc`, and `valloc`. </br>
+  Replacement happens by statically linking the allocator library before the GNU C Library.
+
+- The CCF team has implemented compiling the Open Enclave SDK to use `snmalloc` (Pull Request #2649). Currently, users must recompile OE SDK with `USE_SNMALLOC` to leverage this.
+
+- Rust Language is [considering](https://doc.rust-lang.org/1.15.1/book/custom-allocators.html) capability to switch out the default allocators (`alloc_system` and `alloc_jemalloc`).
+
+
+Non-Requirements
+-----
+
+- This design does not aim to address what the ideal default allocator for Open Enclave should be.
+- Though the design could be extended to support multiple pluggable allocators, each managing a portion of the heap, it is currently a non-requirement.
+
+Requirements
+----
+
+The design for `Pluggable Allocators` additionally has the following requirements,
+
+- Allocators must be pluggable by an SDK user with having to recompile the OE SDK </br>
+  This is similat to the GNU C library in that, one need not rebuild glibc to plugin a malloc replacement.
+
+- Must work on various TEE environments that the Open Enclave SDK will support. </br>
+  Note, it does not mean that a given allocator must work on all platforms; rather Open Enclave SDK must support plugging allocators on all platforms. </br>
+  From a design perspective, this precludes the use of architecture specific datastructures in the plugin interface.
+
+- The design must support allocators to play well with features that Open Enclave SDK already provides (`debug allocators`, `backtracing`, `stack unwinding`, `failure callbacks` etc)
```

Comment:
So these feel like two different things:

* debug allocators
* failure call backs

Are layered on top of the allocator.  While

* backtracing 
* stack unwinding 

Are core platform features that require allocation and often during failure so care is needed.

I would suggest that you actually make `debug allocators` and `failure callbacks` something that the allocator chooses to provide.  Many allocators have considerably more security features than your debug allocator, and by providing them as part of the allocator can do it considerably more efficiently.  For instance, `snmalloc` can detect leaks during teardown with no runtime overhead, by carefully checking each threads data-structures. Similarly for failure call-backs this can be done deep inside the allocator again taking checking off the critical path.

So I would be careful what you layer on top of a high-perf allocator:  `snmalloc`s fast path allocation is ~15 instructions with two not-taken conditional branches.

Reply:
@mjp41 Good point about allocators themselves providing debug/leak detection capabilities. We can extend the contract in next iterations to allow allocators to expose more of their debugging capabilities.

Reply:
@mjp41 Couple of questions about snmalloc
- Does snmalloc pinpoint where the leaks originate by tracking the callstack of alloctions?
- How are you counting the instructions in fast-path? Do you enforce that all the function calls are inlined in the fast path?

Can you provide us with a small sample that can show case snmalloc? I can add that to our samples in one of the subsequent PRs.

Reply:
So at the moment we only use the function to check everything has been deallocated in unit tests:
  https://github.com/microsoft/snmalloc/blob/51e919385fbb30a29e2ab815b9c122e0fbd7cf4b/src/test/func/memory/memory.cc#L115

There are also issues that function that do not deallocate until very late can cause this not to work.  For instance, `at_exit` allocates a block, which is not deallocated until after most stuff is torn down.  But in this case we could put it in the enclave tear down. 

We don't currently track the allocation sites.  ASAN's allocator has some nice features for doing that.  But we have not had the time to do that with `snmalloc`.

Re: Fast path.  `snmalloc` has two macros:
```
SNMALLOC_FAST_PATH
SNMALLOC_SLOW_PATH
```
https://github.com/microsoft/snmalloc/blob/0a081cacd649e138cb83edb56f773c7ecf6b6e2d/src/ds/defines.h#L8-L9

We use this to careful force inlining, and prevent other inlining. A lot of the code is written to cause the compiler to force tail call optimisation:
```C++
   if (likely(condition))
      return fast(size);
   return slow(size);
```
Where `fast` is declared with `SNMALLOC_FAST_PATH` and `slow` with `SNMALLOC_SLOW_PATH`. I am in the process of refactoring the slow paths to be of this form, so we get better codegen for those too.

This then gives a very nice allocation sequence:
```asm
0000000000028930 <malloc>:
   28930:	48 89 fe             	mov    %rdi,%rsi
   28933:	48 8b 05 ce 28 00 00 	mov    0x28ce(%rip),%rax        # 2b208 <_DYNAMIC+0x1e0>
   2893a:	64 48 8b 38          	mov    %fs:(%rax),%rdi
   2893e:	48 8d 46 ff          	lea    -0x1(%rsi),%rax
   28942:	48 3d ff ff 00 00    	cmp    $0xffff,%rax
   28948:	77 20                	ja     2896a <malloc+0x3a>
   2894a:	48 83 e0 f8          	and    $0xfffffffffffffff8,%rax
   2894e:	48 8d 0d 7b 86 fe ff 	lea    -0x17985(%rip),%rcx        # 10fd0 <_ZN8snmallocL18sizeclass_metadataE>
   28955:	48 8b 34 08          	mov    (%rax,%rcx,1),%rsi
   28959:	48 8b 04 f7          	mov    (%rdi,%rsi,8),%rax
   2895d:	48 85 c0             	test   %rax,%rax
   28960:	74 0d                	je     2896f <malloc+0x3f>
   28962:	48 8b 08             	mov    (%rax),%rcx
   28965:	48 89 0c f7          	mov    %rcx,(%rdi,%rsi,8)
   28969:	c3                   	retq   
   2896a:	e9 31 eb ff ff       	jmpq   274a0 <_ZN8snmalloc9AllocatorIXadL_ZNS_20needs_initialisationEPvEEXadL_ZNS_21init_thread_allocatorEvEENS_24MemoryProviderStateMixinINS_8PALLinuxEEENS_15DefaultChunkMapINS_21GlobalPagemapTemplateINS_11FlatPagemapILm24EhEEEEEELb1EE15alloc_not_smallILNS_7ZeroMemE0ELNS_12AllowReserveE1EEES1_m>
   2896f:	48 8b 87 c0 0d 00 00 	mov    0xdc0(%rdi),%rax
   28976:	48 39 87 c8 0d 00 00 	cmp    %rax,0xdc8(%rdi)
   2897d:	75 05                	jne    28984 <malloc+0x54>
   2897f:	e9 2c ec ff ff       	jmpq   275b0 <_ZN8snmalloc9AllocatorIXadL_ZNS_20needs_initialisationEPvEEXadL_ZNS_21init_thread_allocatorEvEENS_24MemoryProviderStateMixinINS_8PALLinuxEEENS_15DefaultChunkMapINS_21GlobalPagemapTemplateINS_11FlatPagemapILm24EhEEEEEELb1EE26small_alloc_next_free_listILNS_7ZeroMemE0ELNS_12AllowReserveE1EEES1_m>
   28984:	e9 c7 ec ff ff       	jmpq   27650 <_ZN8snmalloc9AllocatorIXadL_ZNS_20needs_initialisationEPvEEXadL_ZNS_21init_thread_allocatorEvEENS_24MemoryProviderStateMixinINS_8PALLinuxEEENS_15DefaultChunkMapINS_21GlobalPagemapTemplateINS_11FlatPagemapILm24EhEEEEEELb1EE19small_alloc_mq_slowILNS_7ZeroMemE0ELNS_12AllowReserveE1EEES1_m>
   28989:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)
```

> Can you provide us with a small sample that can show case snmalloc?

I would love to, but I am not entirely sure what you are after. Do you mean allocation patterns that work really well for `snmalloc`?

Reply:
Yes. Allocation patterns would be good.

It would be good to have a sample that demonstrates how an user could plug-in their own allocator. It makes sense to show case `snmalloc` as the custom allocator since it already has a PAL for Open Enclave.

The sample could be as simple as doing a bunch of allocations; however is there something about `snmalloc` that you would want to showcase via a sample?

Reply:
>
> We don't currently track the allocation sites. ASAN's allocator has some nice features for doing that. But we have not had the time to do that with snmalloc.
>

I think it is better that Open Enclave SDK stops at supporting pluggable allocators. It does not seem to add much value to funnel all the debugging capabilities of a particular allocator via some unified debug api in `allocator.h`. Once the application dependent allocator has been plugged in, the use can interface with the debugging capabilities of the allocator directly.

To reiterate, we can revisit user's interaction with debugging capabilities of custom allocators in future iterations.

Note: Currently `debug allocator` which tracks allocation sites is not a public feature. It is only used internally for detecting leaks in the SDK.

Reply:
> Once the application dependent allocator has been plugged in, the use can interface with the debugging capabilities of the allocator directly.

Agreed, I think the point I should have made is: "don't force pluggable allocators to work with the debug allocator in OE, or the failure callbacks."  They can probably do something better if required.

Reply:
> For my understanding, is the conclusion of this discussion that pluggable allocators:
> 
>     * need to provide support for backtracing & stack unwinding
> 
>     * do not need to provide debug allocators and failure callbacks
> 
> 
> For the latter case, given that debug allocator use in OE SDK is internal and not compatible with pluggable allocators, is this an issue for CI/CD configuration in running tests against pluggable allocators?

Here are the answers:
- No allocators do not need to provide support for backtracing and stack unwinding. I will remove that confusing line from the doc.
- Allocators do not need to provide "debug alloctors" and "Failure callbacks". OE will provide these functionality atop the custom allocators, as it currently does atop dlmalloc.

- Debug allocators built atop custom allocators will not cause issues for CI/CD. We do need to lock down the custom allocator support in our CI/CD configuration.